
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://doublehelix.app/advanced-topics/">
      
      
        <link rel="prev" href="../API/">
      
      
        <link rel="next" href="../configuration/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.26">
    
    
      
        <title>Advanced topics - Documentation for DoubleHelix</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#advanced-topics" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Documentation for DoubleHelix" class="md-header__button md-logo" aria-label="Documentation for DoubleHelix" data-md-component="logo">
      
  <img src="../logo.jpg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Documentation for DoubleHelix
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Advanced topics
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/DoubleHelixApp/DoubleHelix" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Documentation for DoubleHelix" class="md-nav__button md-logo" aria-label="Documentation for DoubleHelix" data-md-component="logo">
      
  <img src="../logo.jpg" alt="logo">

    </a>
    Documentation for DoubleHelix
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/DoubleHelixApp/DoubleHelix" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Welcome to DoubleHelix
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../API/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    API
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Advanced topics
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Advanced topics
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#reference-genome" class="md-nav__link">
    <span class="md-ellipsis">
      Reference genome
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Reference genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-it-works" class="md-nav__link">
    <span class="md-ellipsis">
      How it works
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-a-new-reference" class="md-nav__link">
    <span class="md-ellipsis">
      Add a new reference
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#progress-bar" class="md-nav__link">
    <span class="md-ellipsis">
      Progress bar
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data folder
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../glossary/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Glossary
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../user-guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    User guide
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#reference-genome" class="md-nav__link">
    <span class="md-ellipsis">
      Reference genome
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Reference genome">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-it-works" class="md-nav__link">
    <span class="md-ellipsis">
      How it works
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-a-new-reference" class="md-nav__link">
    <span class="md-ellipsis">
      Add a new reference
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#progress-bar" class="md-nav__link">
    <span class="md-ellipsis">
      Progress bar
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="advanced-topics">Advanced topics</h1>
<h2 id="reference-genome">Reference genome</h2>
<p>In some cases, DoubleHelix needs to access the reference genome that was used to create an alignment-map file. Since this information is not contained in an alignment-map file, DoubleHelix needs to guess it. This document explains how DoubleHelix solves this problem and what can go wrong.</p>
<h3 id="introduction">Introduction</h3>
<p>The biggest hint for identifying the reference genome is the header of an alignment-map file. The header contains a list of sequences contained in the reference used to do the alignment.</p>
<p>DoubleHelix solves the issue of identifying the reference by maintaining a list of meta-information about several reference genomes. In this way, it can identify a reference if it's present in this list. The list can be modified and new references can be suggested by using this <a href="https://github.com/DoubleHelixApp/DoubleHelix/issues/new?assignees=chaplin89&amp;labels=reference&amp;projects=&amp;template=add-a-new-reference.md&amp;title=%5BReference%5D+Please+add+a+new+reference">GitHub issue template</a> or by submitting a PR using the instruction below.</p>
<p>The process is not perfect and it can potentially give incorrect information (see the section below to understand how it works). In this case, some actions may fail. Please feel free to open a <a href="https://github.com/DoubleHelixApp/DoubleHelix/issues/new?assignees=&amp;labels=&amp;projects=&amp;template=bug_report.md&amp;title=">bug report</a> if that happens.</p>
<h3 id="how-it-works">How it works</h3>
<p>The identification can work in two different ways, depending on the information available in the alignment-map file: </p>
<ul>
<li>Using MD5 of sequences </li>
<li>Using lengths of sequences</li>
</ul>
<p>MD5 indicates an MD5 string calculated in a reliable way (that accounts for case differences inside the sequence) over the content of a sequence. The procedure to calculate the MD5 is described on page 6 of the SAM standard. An MD5 can identify unequivocally a specific sequence and it's the preferred method to find a reference. If MD5 of the reference sequences are available it's easy to find the reference looking in the meta-data list of reference genomes that DoubleHelix maintains. Lengths are used only as a fallback when MD5s are not available. Unfortunately, this situation is the most common one. Despite MD5 sequences being part of the SAM specification standard they are not mandatory. Most alignment-map files won't have this field populated.</p>
<p>The lengths indicate the length of the sequences expressed in base pairs. A length itself cannot reliably identify a sequence as it's possible (and happening in practices) to have the same lengths for sequences having completely different content (and hence a different MD5). What's more reliable is to use the whole set of sequences to identify a reference. The sequence of lengths contained in a reference is pretty unique for each reference, and it's the current way used by DoubleHelix to identify a reference when the MD5 is not available. It's still possible that two references have a perfectly identical set of lengths having a different sequence of MD5s. In this case, if the alignment-map file was associated by DoubleHelix to a reference falling in this situation, DoubleHelix will present a choice to the user. Since it's impossible to reliably determine which reference was used without other information, the user has to choose which reference is the correct one.</p>
<p>This is a list of ambiguous lengths sequences (will be updated at  every commit)</p>
<ol>
<li>https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/analysisSet/hg38.analysisSet.fa.gz</li>
<li>
<p>https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Eukaryotes/vertebrates_mammals/Homo_sapiens/GRCh38/seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz</p>
</li>
<li>
<p>https://ftp.ensembl.org/pub/release-55/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.55.dna.toplevel.fa.gz</p>
</li>
<li>https://ftp.ensembl.org/pub/release-56/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.56.dna.toplevel.fa.gz</li>
</ol>
<h3 id="add-a-new-reference">Add a new reference</h3>
<p>It's possible to onboard a new reference to DoubleHelix either by opening a GitHub issue with <a href="https://github.com/DoubleHelixApp/DoubleHelix/issues/new?assignees=chaplin89&amp;labels=reference&amp;projects=&amp;template=add-a-new-reference.md&amp;title=%5BReference%5D+Please+add+a+new+reference">this template</a> or by using the following code:</p>
<div class="highlight"><pre><span></span><code><span class="n">manager</span> <span class="o">=</span> <span class="n">RepositoryManager</span><span class="p">()</span>
<span class="n">manager</span><span class="o">.</span><span class="n">genomes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
    <span class="n">manager</span><span class="o">.</span><span class="n">ingest</span><span class="p">(</span>
        <span class="s2">&quot;https://source/reference.fa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NIH&quot;</span><span class="p">,</span> <span class="c1"># Anything that matches an entry in sources.json, otherwise add an entry there</span>
        <span class="s2">&quot;38&quot;</span><span class="p">,</span>  <span class="c1"># Only 38 or 19</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">GenomeMetadataLoader</span><span class="p">()</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">manager</span><span class="o">.</span><span class="n">genomes</span><span class="p">)</span>
</code></pre></div>
<p>A GUI/CLI way will be added in the future.</p>
<h2 id="progress-bar">Progress bar</h2>
<p>To provide a friendlier interface, DoubleHelix shows a progress bar while doing long operations. Implementing a progress bar for DoubleHelix required a considerable effort, as DoubleHelix is reliant on dependencies who doesn't allow to monitor their progress in any way.</p>
<p>In these cases, the approach DoubleHelix is following is to estimate how many bytes will the process needs to read or write and constantly polling the current amount of bytes processed to understand where the process is. Is not easy to estimate this info in some cases, and in general the process is quite hacky, meaning it can break easily (giving incorrect ETA or percentage) after advancing the version of these dependencies. Some classes to abstract this behaviour are making this process as painless as possible.</p>
<p>The architecture like this: ProcessIOMonitor -&gt; BaseProgressCalculator -&gt; UI Callback</p>
<p>All these classes are in the <code>helix.progress</code> namespace:</p>
<ul>
<li><em>helix.progress.ProcessIOMonitor</em>: will start a process and preiodically calling a callback, supplying <a href="https://psutil.readthedocs.io/en/latest/#psutil.disk_io_counters"><code>io_counters</code></a> (i.e., read/write bytes from the process).</li>
<li><em>helix.progress.BaseProcessCalculator</em>: Implements the callback for the class above. <code>BaseProgressCalculator</code> knows what's the target to reach and convert the bytes supplied by <code>ProcessIOMonitor</code> into a percentage and an ETA. It implements all the features needed to have stable values.</li>
<li><em>helix.progress.FileIOMonitor</em>: Is the equivalent of ProcessIOMonitor but it monitors read or write bytes on a file, not on the process itself. Sometime is simply more practical to monitor the growth of a file, as a process may not exists (i.e., the file is generate by another 3rd party python module that does not provide any way to monitor the progress). It provides an interface that is similar to <code>ProcessIOMonitor</code>.</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
    
  </body>
</html>