
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://doublehelix.app/API/">
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../advanced-topics/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.26">
    
    
      
        <title>API - Documentation for DoubleHelix</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Documentation for DoubleHelix" class="md-header__button md-logo" aria-label="Documentation for DoubleHelix" data-md-component="logo">
      
  <img src="../logo.jpg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Documentation for DoubleHelix
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              API
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/DoubleHelixApp/DoubleHelix" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Documentation for DoubleHelix" class="md-nav__button md-logo" aria-label="Documentation for DoubleHelix" data-md-component="logo">
      
  <img src="../logo.jpg" alt="logo">

    </a>
    Documentation for DoubleHelix
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/DoubleHelixApp/DoubleHelix" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Welcome to DoubleHelix
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    API
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    API
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#repository-class" class="md-nav__link">
    <span class="md-ellipsis">
      Repository class
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository" class="md-nav__link">
    <span class="md-ellipsis">
      Repository
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Repository">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository--attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Attributes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository--methods" class="md-nav__link">
    <span class="md-ellipsis">
      Methods
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository--insert-docstring-explaining-methods-if-any" class="md-nav__link">
    <span class="md-ellipsis">
      Insert docstring explaining methods, if any.
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.analyze_references" class="md-nav__link">
    <span class="md-ellipsis">
      analyze_references
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.download" class="md-nav__link">
    <span class="md-ellipsis">
      download
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.find" class="md-nav__link">
    <span class="md-ellipsis">
      find
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.ingest" class="md-nav__link">
    <span class="md-ellipsis">
      ingest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.self_test" class="md-nav__link">
    <span class="md-ellipsis">
      self_test
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#unitprefix-class" class="md-nav__link">
    <span class="md-ellipsis">
      UnitPrefix class
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.utility.unit_prefix.UnitPrefix" class="md-nav__link">
    <span class="md-ellipsis">
      UnitPrefix
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.utility.unit_prefix.UnitPrefix.convert" class="md-nav__link">
    <span class="md-ellipsis">
      convert
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.utility.unit_prefix.UnitPrefix.convert_bytes" class="md-nav__link">
    <span class="md-ellipsis">
      convert_bytes
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../advanced-topics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Advanced topics
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../user-guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    User guide
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#repository-class" class="md-nav__link">
    <span class="md-ellipsis">
      Repository class
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository" class="md-nav__link">
    <span class="md-ellipsis">
      Repository
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Repository">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository--attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Attributes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository--methods" class="md-nav__link">
    <span class="md-ellipsis">
      Methods
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository--insert-docstring-explaining-methods-if-any" class="md-nav__link">
    <span class="md-ellipsis">
      Insert docstring explaining methods, if any.
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.analyze_references" class="md-nav__link">
    <span class="md-ellipsis">
      analyze_references
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.download" class="md-nav__link">
    <span class="md-ellipsis">
      download
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.find" class="md-nav__link">
    <span class="md-ellipsis">
      find
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.ingest" class="md-nav__link">
    <span class="md-ellipsis">
      ingest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.reference.repository_manager.Repository.self_test" class="md-nav__link">
    <span class="md-ellipsis">
      self_test
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#unitprefix-class" class="md-nav__link">
    <span class="md-ellipsis">
      UnitPrefix class
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.utility.unit_prefix.UnitPrefix" class="md-nav__link">
    <span class="md-ellipsis">
      UnitPrefix
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.utility.unit_prefix.UnitPrefix.convert" class="md-nav__link">
    <span class="md-ellipsis">
      convert
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#helix.utility.unit_prefix.UnitPrefix.convert_bytes" class="md-nav__link">
    <span class="md-ellipsis">
      convert_bytes
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="api">API</h1>
<h2 id="repository-class">Repository class</h2>


<div class="doc doc-object doc-class">



<a id="helix.reference.repository_manager.Repository"></a>
    <div class="doc doc-contents first">


      <p>The central repository class that holds all the necessary objects to perform
various operations. This class encapsulates a set of functionalities that are
related to biological data processing and manipulation.</p>
<h3 id="helix.reference.repository_manager.Repository--attributes">Attributes</h3>
<pre><code>* alignment_map_header: AlignmentMapHeader
    - A header object representing an alignment map.
* configuration: RepositoryConfig
    - The repository's configuration settings.
* decompressor: Decompressor
    - An object for decompressing files.
* downloader: Downloader
    - A class used to download files from a remote location.
* genome: Genome
    - Represents the genomic data.
* metadata_loader: MetadataLoader
    - Used to load metadata about genomes and sequences.
* mt_dna: MtDNA
    - Handles mitochondrial DNA-related operations.
* samtools: Samtools
    - An object providing functions for manipulating SAM/BAM files.
</code></pre>
<h3 id="helix.reference.repository_manager.Repository--methods">Methods</h3>
<h2 id="helix.reference.repository_manager.Repository--insert-docstring-explaining-methods-if-any">Insert docstring explaining methods, if any.</h2>

              <details class="quote">
                <summary>Source code in <code>helix\reference\repository_manager.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Repository</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The central repository class that holds all the necessary objects to perform</span>
<span class="sd">    various operations. This class encapsulates a set of functionalities that are</span>
<span class="sd">    related to biological data processing and manipulation.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>

<span class="sd">        * alignment_map_header: AlignmentMapHeader</span>
<span class="sd">            - A header object representing an alignment map.</span>
<span class="sd">        * configuration: RepositoryConfig</span>
<span class="sd">            - The repository&#39;s configuration settings.</span>
<span class="sd">        * decompressor: Decompressor</span>
<span class="sd">            - An object for decompressing files.</span>
<span class="sd">        * downloader: Downloader</span>
<span class="sd">            - A class used to download files from a remote location.</span>
<span class="sd">        * genome: Genome</span>
<span class="sd">            - Represents the genomic data.</span>
<span class="sd">        * metadata_loader: MetadataLoader</span>
<span class="sd">            - Used to load metadata about genomes and sequences.</span>
<span class="sd">        * mt_dna: MtDNA</span>
<span class="sd">            - Handles mitochondrial DNA-related operations.</span>
<span class="sd">        * samtools: Samtools</span>
<span class="sd">            - An object providing functions for manipulating SAM/BAM files.</span>

<span class="sd">    Methods</span>
<span class="sd">    -------</span>

<span class="sd">    # Insert docstring explaining methods, if any.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">loader</span><span class="p">:</span> <span class="n">MetadataLoader</span> <span class="o">=</span> <span class="n">MetadataLoader</span><span class="p">(),</span>
        <span class="n">downloader</span><span class="p">:</span> <span class="n">Downloader</span> <span class="o">=</span> <span class="n">Downloader</span><span class="p">(),</span>
        <span class="n">mtdna</span><span class="p">:</span> <span class="n">MtDNA</span> <span class="o">=</span> <span class="n">MtDNA</span><span class="p">(),</span>
        <span class="n">config</span><span class="p">:</span> <span class="n">RepositoryConfig</span> <span class="o">=</span> <span class="n">RepositoryConfig</span><span class="p">(),</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_loader</span> <span class="o">=</span> <span class="n">loader</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_downloader</span> <span class="o">=</span> <span class="n">downloader</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mtdna</span> <span class="o">=</span> <span class="n">mtdna</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_loader</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
        <span class="c1"># Index individual sequences by their MD5 and length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_md5</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_group_sequences</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">md5</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_group_sequences</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_config</span> <span class="o">=</span> <span class="n">config</span>

    <span class="k">def</span> <span class="nf">analyze_references</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Analyze the references in the repository and provide some basic information.</span>

<span class="sd">        This method currently just prints out some statistics about the sequences.</span>
<span class="sd">        It needs to be improved to provide more useful information about supported</span>
<span class="sd">        references. For now, it&#39;s just a placeholder.</span>

<span class="sd">        :return: None</span>
<span class="sd">        :rtype: NoneType</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Analyzing sequence by sequence</span>
        <span class="n">different_md5_same_length</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">length</span><span class="p">,</span> <span class="n">seqs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">unique_seq_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">:</span>
                <span class="n">unique_seq_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">md5</span><span class="p">)</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">name</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_seq_set</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">57227414</span><span class="p">:</span>
                <span class="n">different_md5_same_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_seq_set</span><span class="p">)</span><span class="si">}</span><span class="s2"> different MD5 for length </span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">).&quot;</span>
                <span class="p">)</span>

        <span class="c1"># Analyzing the whole genome, looking for identical length sequences</span>
        <span class="c1"># with different MD5 sequences</span>
        <span class="n">different_md5s_same_lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">seq_lengths</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">length</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
            <span class="n">seq_md5</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">md5</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">seq_lengths</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="p">:</span>
                <span class="n">different_md5s_same_lengths</span><span class="p">[</span><span class="n">seq_lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">seq_md5</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="p">[</span><span class="n">seq_lengths</span><span class="p">]]:</span>
                <span class="n">different_md5s_same_lengths</span><span class="p">[</span><span class="n">seq_lengths</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">seq_md5</span><span class="p">,</span> <span class="n">genome</span><span class="p">))</span>

        <span class="n">ambiguous</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span>
            <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">ambiguous</span><span class="si">}</span><span class="s2"> possible ambiguous sequences of lengths:&quot;</span><span class="p">)</span>
        <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">length</span><span class="p">,</span> <span class="n">md5s</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">md5s</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">md5</span><span class="p">,</span> <span class="n">genome</span> <span class="ow">in</span> <span class="n">md5s</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">) </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span><span class="si">!s}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">def</span> <span class="nf">find</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequences</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find the references that match the given sequence(s) based on their MD5</span>
<span class="sd">        or length if an MD5 is not available.</span>

<span class="sd">        This method returns a Reference object containing the matching sequences.</span>

<span class="sd">        :param sequences: A list of Sequence objects to search for.</span>
<span class="sd">        :return: A Reference object with the matching sequences.</span>
<span class="sd">        :rtype: helix.reference.reference.Reference</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">md5_available</span> <span class="o">=</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">md5</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">])</span>
        <span class="n">matching</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">([(</span><span class="n">x</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">md5_available</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">md5</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_md5</span><span class="p">:</span>
                    <span class="n">matching</span><span class="p">[</span><span class="n">sequence</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_md5</span><span class="p">[</span><span class="n">sequence</span><span class="o">.</span><span class="n">md5</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">length</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="p">:</span>
                    <span class="n">matching</span><span class="p">[</span><span class="n">sequence</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="p">[</span><span class="n">sequence</span><span class="o">.</span><span class="n">length</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">Reference</span><span class="p">(</span><span class="n">matching</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_group_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">criteria</span><span class="p">,</span> <span class="n">genomes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Genome</span><span class="p">]):</span>
        <span class="n">grouped</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="n">genomes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
                <span class="n">evaluated_criteria</span> <span class="o">=</span> <span class="n">criteria</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">evaluated_criteria</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">grouped</span><span class="p">:</span>
                    <span class="n">grouped</span><span class="p">[</span><span class="n">evaluated_criteria</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">grouped</span><span class="p">[</span><span class="n">evaluated_criteria</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">grouped</span>

    <span class="k">def</span> <span class="nf">_get_sizes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Genome</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">fai_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fai_url</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span> <span class="o">+</span> <span class="s2">&quot;.fai&quot;</span>
            <span class="n">fai_size</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_downloader</span><span class="o">.</span><span class="n">get_file_size</span><span class="p">(</span><span class="n">fai_url</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">fai_size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">fai_url</span> <span class="o">=</span> <span class="n">fai_url</span>

        <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">gzi_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gzi_url</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span> <span class="o">+</span> <span class="s2">&quot;.gzi&quot;</span>
            <span class="n">gzi_size</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_downloader</span><span class="o">.</span><span class="n">get_file_size</span><span class="p">(</span><span class="n">gzi_url</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gzi_size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">gzi_url</span> <span class="o">=</span> <span class="n">gzi_url</span>

        <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">download_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">size</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_downloader</span><span class="o">.</span><span class="n">get_file_size</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">download_size</span> <span class="o">=</span> <span class="n">size</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unable to get the size of the fasta file for </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Genome</span><span class="p">):</span>
        <span class="n">dictionary</span><span class="p">:</span> <span class="n">AlignmentMapHeader</span> <span class="o">=</span> <span class="n">AlignmentMapHeader</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">dict</span><span class="p">)</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">dictionary</span><span class="o">.</span><span class="n">sequences</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">entry</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">entry</span><span class="o">.</span><span class="n">md5</span><span class="p">)</span>
            <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sequences</span>

    <span class="k">def</span> <span class="nf">_create_companion_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="n">samtools</span> <span class="o">=</span> <span class="n">Samtools</span><span class="p">()</span>
        <span class="n">bgzip</span> <span class="o">=</span> <span class="n">BGzip</span><span class="p">()</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">: Starting post-download tasks.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome</span><span class="o">.</span><span class="n">gzi</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">: Generating bgzip index.&quot;</span><span class="p">)</span>
            <span class="n">bgzip</span><span class="o">.</span><span class="n">bgzip_wrapper</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="p">,</span> <span class="n">genome</span><span class="o">.</span><span class="n">gzi</span><span class="p">,</span> <span class="n">BgzipAction</span><span class="o">.</span><span class="n">Reindex</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">: bgzip index exists.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome</span><span class="o">.</span><span class="n">dict</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">: Generating dictionary file.&quot;</span><span class="p">)</span>
            <span class="n">samtools</span><span class="o">.</span><span class="n">make_dictionary</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="p">,</span> <span class="n">genome</span><span class="o">.</span><span class="n">dict</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">: Dictionary file exists.&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">self_test</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Runs tests on the current reference data to ensure its integrity and consistency.</span>

<span class="sd">        This method checks if all genomes have sequences by comparing the lengths of their</span>
<span class="sd">        sequence lists with the number of sequences in their fasta files. If any genome is found</span>
<span class="sd">        without sequences, it attempts to download the missing sequences from the original source.</span>

<span class="sd">        :return: None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">reference</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">reference</span><span class="o">.</span><span class="n">sequences</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="n">reference</span><span class="si">}</span><span class="s2"> as it has no sequences.&quot;</span><span class="p">)</span>
                    <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ingest</span><span class="p">(</span>
                        <span class="n">reference</span><span class="o">.</span><span class="n">fasta_url</span><span class="p">,</span> <span class="n">reference</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">reference</span><span class="o">.</span><span class="n">build</span>
                    <span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_loader</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">download</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">progress</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This Python function downloads, decompresses, compresses, and extracts sequence data from a</span>
<span class="sd">        genome file, with optional progress tracking and force download capability.</span>

<span class="sd">        :param genome: Represents the genome for which the download operation is being performed</span>
<span class="sd">        :type genome: Genome</span>
<span class="sd">        :param progress: Callback function that takes two arguments: a string indicating the progress</span>
<span class="sd">        message and an integer indicating the progress percentage.</span>
<span class="sd">        :type progress: Callable[[str, int], None]</span>
<span class="sd">        :param force: Determines whether to overwrite existing files. If `force` is set to `True`, the</span>
<span class="sd">        method will delete any existing files related to the genome being downloaded and re-download them.</span>
<span class="sd">        :return: The `download` method returns the `genome` object after performing the download,</span>
<span class="sd">        decompression, compression, and creation of companion files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">downloader</span> <span class="o">=</span> <span class="n">Downloader</span><span class="p">()</span>
        <span class="n">decompressor</span> <span class="o">=</span> <span class="n">Decompressor</span><span class="p">()</span>
        <span class="n">compressor</span> <span class="o">=</span> <span class="n">BGzip</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">force</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> already exist. Re-using it.&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_create_companion_files</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">force</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">genome</span>
        <span class="k">elif</span> <span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="n">force</span><span class="p">:</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">bgzip_md5</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">downloaded_md5</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">decompressed_md5</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">bgzip_size</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">download_size</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">decompressed_size</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;1/4: Downloading fasta from: </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">download_output</span> <span class="o">=</span> <span class="n">downloader</span><span class="o">.</span><span class="n">perform</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">progress</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;2/4: Decompressing: </span><span class="si">{</span><span class="n">download_output</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">decompressor_output</span> <span class="o">=</span> <span class="n">decompressor</span><span class="o">.</span><span class="n">perform</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">download_output</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;3/4: Compressing to gzip: </span><span class="si">{</span><span class="n">decompressor_output</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">compressor_output</span> <span class="o">=</span> <span class="n">compressor</span><span class="o">.</span><span class="n">perform</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">decompressor_output</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;4/4: Creating companion files: </span><span class="si">{</span><span class="n">compressor_output</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_companion_files</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">genome</span>

    <span class="k">def</span> <span class="nf">ingest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a new Genome object and ingests the provided URL, source, and build information.</span>

<span class="sd">        This method downloads, decompresses, compresses, and extracts sequence data from a</span>
<span class="sd">        genome file, with optional progress tracking and force download capability. The created</span>
<span class="sd">        Genome object is then returned.</span>

<span class="sd">        :param url: The URL of the genome file</span>
<span class="sd">        :type url: str</span>
<span class="sd">        :param source: The source of the genome file</span>
<span class="sd">        :type source: str</span>
<span class="sd">        :param build: The build version of the genome file</span>
<span class="sd">        :type build: str</span>
<span class="sd">        :param force: Determines whether to overwrite existing files. If `force` is set to `True`, the</span>
<span class="sd">        method will delete any existing files related to the genome being downloaded and re-download them.</span>
<span class="sd">        :return: A new Genome object after performing the download, decompression, compression, and creation of companion files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="n">build</span><span class="p">)</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">parent_folder</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="o">.</span><span class="n">genomes</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ingesting </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">force</span><span class="p">)</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_sequences</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">genome</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="helix.reference.repository_manager.Repository.analyze_references" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">analyze_references</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Analyze the references in the repository and provide some basic information.</p>
<p>This method currently just prints out some statistics about the sequences.
It needs to be improved to provide more useful information about supported
references. For now, it's just a placeholder.</p>
<p>:return: None
:rtype: NoneType</p>

            <details class="quote">
              <summary>Source code in <code>helix\reference\repository_manager.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">analyze_references</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyze the references in the repository and provide some basic information.</span>

<span class="sd">    This method currently just prints out some statistics about the sequences.</span>
<span class="sd">    It needs to be improved to provide more useful information about supported</span>
<span class="sd">    references. For now, it&#39;s just a placeholder.</span>

<span class="sd">    :return: None</span>
<span class="sd">    :rtype: NoneType</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Analyzing sequence by sequence</span>
    <span class="n">different_md5_same_length</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">length</span><span class="p">,</span> <span class="n">seqs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">unique_seq_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">:</span>
            <span class="n">unique_seq_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">md5</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">name</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_seq_set</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">57227414</span><span class="p">:</span>
            <span class="n">different_md5_same_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">length</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;There are </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_seq_set</span><span class="p">)</span><span class="si">}</span><span class="s2"> different MD5 for length </span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">).&quot;</span>
            <span class="p">)</span>

    <span class="c1"># Analyzing the whole genome, looking for identical length sequences</span>
    <span class="c1"># with different MD5 sequences</span>
    <span class="n">different_md5s_same_lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">seq_lengths</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">length</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
        <span class="n">seq_md5</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">md5</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">seq_lengths</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="p">:</span>
            <span class="n">different_md5s_same_lengths</span><span class="p">[</span><span class="n">seq_lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">seq_md5</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="p">[</span><span class="n">seq_lengths</span><span class="p">]]:</span>
            <span class="n">different_md5s_same_lengths</span><span class="p">[</span><span class="n">seq_lengths</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">seq_md5</span><span class="p">,</span> <span class="n">genome</span><span class="p">))</span>

    <span class="n">ambiguous</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span>
        <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">ambiguous</span><span class="si">}</span><span class="s2"> possible ambiguous sequences of lengths:&quot;</span><span class="p">)</span>
    <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">length</span><span class="p">,</span> <span class="n">md5s</span> <span class="ow">in</span> <span class="n">different_md5s_same_lengths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">md5s</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">md5</span><span class="p">,</span> <span class="n">genome</span> <span class="ow">in</span> <span class="n">md5s</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">) </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span><span class="si">!s}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="helix.reference.repository_manager.Repository.download" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">download</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">progress</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>This Python function downloads, decompresses, compresses, and extracts sequence data from a
genome file, with optional progress tracking and force download capability.</p>
<p>:param genome: Represents the genome for which the download operation is being performed
:type genome: Genome
:param progress: Callback function that takes two arguments: a string indicating the progress
message and an integer indicating the progress percentage.
:type progress: Callable[[str, int], None]
:param force: Determines whether to overwrite existing files. If <code>force</code> is set to <code>True</code>, the
method will delete any existing files related to the genome being downloaded and re-download them.
:return: The <code>download</code> method returns the <code>genome</code> object after performing the download,
decompression, compression, and creation of companion files.</p>

            <details class="quote">
              <summary>Source code in <code>helix\reference\repository_manager.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">download</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">progress</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This Python function downloads, decompresses, compresses, and extracts sequence data from a</span>
<span class="sd">    genome file, with optional progress tracking and force download capability.</span>

<span class="sd">    :param genome: Represents the genome for which the download operation is being performed</span>
<span class="sd">    :type genome: Genome</span>
<span class="sd">    :param progress: Callback function that takes two arguments: a string indicating the progress</span>
<span class="sd">    message and an integer indicating the progress percentage.</span>
<span class="sd">    :type progress: Callable[[str, int], None]</span>
<span class="sd">    :param force: Determines whether to overwrite existing files. If `force` is set to `True`, the</span>
<span class="sd">    method will delete any existing files related to the genome being downloaded and re-download them.</span>
<span class="sd">    :return: The `download` method returns the `genome` object after performing the download,</span>
<span class="sd">    decompression, compression, and creation of companion files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">downloader</span> <span class="o">=</span> <span class="n">Downloader</span><span class="p">()</span>
    <span class="n">decompressor</span> <span class="o">=</span> <span class="n">Decompressor</span><span class="p">()</span>
    <span class="n">compressor</span> <span class="o">=</span> <span class="n">BGzip</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">force</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> already exist. Re-using it.&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_companion_files</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">force</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">genome</span>
    <span class="k">elif</span> <span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="n">force</span><span class="p">:</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">fasta</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">bgzip_md5</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">downloaded_md5</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">decompressed_md5</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">bgzip_size</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">download_size</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">decompressed_size</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;1/4: Downloading fasta from: </span><span class="si">{</span><span class="n">genome</span><span class="o">.</span><span class="n">fasta_url</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="n">download_output</span> <span class="o">=</span> <span class="n">downloader</span><span class="o">.</span><span class="n">perform</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">progress</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;2/4: Decompressing: </span><span class="si">{</span><span class="n">download_output</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="n">decompressor_output</span> <span class="o">=</span> <span class="n">decompressor</span><span class="o">.</span><span class="n">perform</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">download_output</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;3/4: Compressing to gzip: </span><span class="si">{</span><span class="n">decompressor_output</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="n">compressor_output</span> <span class="o">=</span> <span class="n">compressor</span><span class="o">.</span><span class="n">perform</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">decompressor_output</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;4/4: Creating companion files: </span><span class="si">{</span><span class="n">compressor_output</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_create_companion_files</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genome</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="helix.reference.repository_manager.Repository.find" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Find the references that match the given sequence(s) based on their MD5
or length if an MD5 is not available.</p>
<p>This method returns a Reference object containing the matching sequences.</p>
<p>:param sequences: A list of Sequence objects to search for.
:return: A Reference object with the matching sequences.
:rtype: helix.reference.reference.Reference</p>

            <details class="quote">
              <summary>Source code in <code>helix\reference\repository_manager.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequences</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Sequence</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the references that match the given sequence(s) based on their MD5</span>
<span class="sd">    or length if an MD5 is not available.</span>

<span class="sd">    This method returns a Reference object containing the matching sequences.</span>

<span class="sd">    :param sequences: A list of Sequence objects to search for.</span>
<span class="sd">    :return: A Reference object with the matching sequences.</span>
<span class="sd">    :rtype: helix.reference.reference.Reference</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">md5_available</span> <span class="o">=</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">md5</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">])</span>
    <span class="n">matching</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">([(</span><span class="n">x</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">md5_available</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">md5</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_md5</span><span class="p">:</span>
                <span class="n">matching</span><span class="p">[</span><span class="n">sequence</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_md5</span><span class="p">[</span><span class="n">sequence</span><span class="o">.</span><span class="n">md5</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">length</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="p">:</span>
                <span class="n">matching</span><span class="p">[</span><span class="n">sequence</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sequences_by_length</span><span class="p">[</span><span class="n">sequence</span><span class="o">.</span><span class="n">length</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">Reference</span><span class="p">(</span><span class="n">matching</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="helix.reference.repository_manager.Repository.ingest" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ingest</span><span class="p">(</span><span class="n">url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Creates a new Genome object and ingests the provided URL, source, and build information.</p>
<p>This method downloads, decompresses, compresses, and extracts sequence data from a
genome file, with optional progress tracking and force download capability. The created
Genome object is then returned.</p>
<p>:param url: The URL of the genome file
:type url: str
:param source: The source of the genome file
:type source: str
:param build: The build version of the genome file
:type build: str
:param force: Determines whether to overwrite existing files. If <code>force</code> is set to <code>True</code>, the
method will delete any existing files related to the genome being downloaded and re-download them.
:return: A new Genome object after performing the download, decompression, compression, and creation of companion files.</p>

            <details class="quote">
              <summary>Source code in <code>helix\reference\repository_manager.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">ingest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates a new Genome object and ingests the provided URL, source, and build information.</span>

<span class="sd">    This method downloads, decompresses, compresses, and extracts sequence data from a</span>
<span class="sd">    genome file, with optional progress tracking and force download capability. The created</span>
<span class="sd">    Genome object is then returned.</span>

<span class="sd">    :param url: The URL of the genome file</span>
<span class="sd">    :type url: str</span>
<span class="sd">    :param source: The source of the genome file</span>
<span class="sd">    :type source: str</span>
<span class="sd">    :param build: The build version of the genome file</span>
<span class="sd">    :type build: str</span>
<span class="sd">    :param force: Determines whether to overwrite existing files. If `force` is set to `True`, the</span>
<span class="sd">    method will delete any existing files related to the genome being downloaded and re-download them.</span>
<span class="sd">    :return: A new Genome object after performing the download, decompression, compression, and creation of companion files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="n">build</span><span class="p">)</span>
    <span class="n">genome</span><span class="o">.</span><span class="n">parent_folder</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_config</span><span class="o">.</span><span class="n">genomes</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ingesting </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">force</span><span class="p">)</span>
    <span class="n">genome</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_sequences</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genome</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="helix.reference.repository_manager.Repository.self_test" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">self_test</span><span class="p">()</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Runs tests on the current reference data to ensure its integrity and consistency.</p>
<p>This method checks if all genomes have sequences by comparing the lengths of their
sequence lists with the number of sequences in their fasta files. If any genome is found
without sequences, it attempts to download the missing sequences from the original source.</p>
<p>:return: None</p>

            <details class="quote">
              <summary>Source code in <code>helix\reference\repository_manager.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">self_test</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Runs tests on the current reference data to ensure its integrity and consistency.</span>

<span class="sd">    This method checks if all genomes have sequences by comparing the lengths of their</span>
<span class="sd">    sequence lists with the number of sequences in their fasta files. If any genome is found</span>
<span class="sd">    without sequences, it attempts to download the missing sequences from the original source.</span>

<span class="sd">    :return: None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">reference</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">reference</span><span class="o">.</span><span class="n">sequences</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="n">reference</span><span class="si">}</span><span class="s2"> as it has no sequences.&quot;</span><span class="p">)</span>
                <span class="n">genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ingest</span><span class="p">(</span>
                    <span class="n">reference</span><span class="o">.</span><span class="n">fasta_url</span><span class="p">,</span> <span class="n">reference</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">reference</span><span class="o">.</span><span class="n">build</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_loader</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomes</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="unitprefix-class">UnitPrefix class</h2>


<div class="doc doc-object doc-class">



<a id="helix.utility.unit_prefix.UnitPrefix"></a>
    <div class="doc doc-contents first">


      <p>Convert a quantity into a more readable form using unit prefixes.</p>
<p>This class is expecially useful for very big or very small number, that
are difficult to read or visualize.</p>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">UnitPrefix</span><span class="o">.</span><span class="n">convert_bytes</span><span class="p">(</span><span class="mi">1024</span><span class="p">)</span>
<span class="go">&#39;1 kB&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnitPrefix</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span>
<span class="go">&#39;1 k&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnitPrefix</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="mf">0.01</span><span class="p">)</span>
<span class="go">&#39;10 m&#39;</span>
</code></pre></div>

              <details class="quote">
                <summary>Source code in <code>helix\utility\unit_prefix.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">UnitPrefix</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a quantity into a more readable form using unit prefixes.</span>

<span class="sd">    This class is expecially useful for very big or very small number, that</span>
<span class="sd">    are difficult to read or visualize.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; UnitPrefix.convert_bytes(1024)</span>
<span class="sd">        &#39;1 kB&#39;</span>
<span class="sd">        &gt;&gt;&gt; UnitPrefix.convert(1000)</span>
<span class="sd">        &#39;1 k&#39;</span>
<span class="sd">        &gt;&gt;&gt; UnitPrefix.convert(0.01)</span>
<span class="sd">        &#39;10 m&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">convert_bytes</span><span class="p">(</span><span class="nb">input</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert bytes using binary prefixes.</span>

<span class="sd">        Args:</span>
<span class="sd">            input (int): The number to convert.</span>
<span class="sd">            decimal (int, optional): Number of decimals to use. Defaults to 2.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: The input number is a float</span>
<span class="sd">            ValueError: The input number is negative.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Converted number.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">input</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Can&#39;t convert a fractional number of bytes&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">input</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Can&#39;t convert a negative number of bytes&quot;</span><span class="p">)</span>
        <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">si_prefixes</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot; B&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">0</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; kB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; MB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">20</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; GB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">30</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; TB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mi">40</span><span class="p">)),</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">input</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;0 B&quot;</span>

        <span class="n">input_log</span> <span class="o">=</span> <span class="n">log</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">input_log</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">input_log</span><span class="p">))</span>

        <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">letter</span><span class="p">,</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">si_prefixes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">min_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">max_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="n">max_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span>
            <span class="k">elif</span> <span class="n">min_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span> <span class="ow">and</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>

            <span class="k">if</span> <span class="n">in_interval</span><span class="p">:</span>
                <span class="n">converted</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">(</span><span class="nb">input</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">2</span><span class="o">**</span><span class="n">interval</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">:</span><span class="s2">.</span><span class="si">{</span><span class="n">decimal</span><span class="si">}</span><span class="s2">f</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="c1"># Not interested in any .0000 at the end</span>
                <span class="n">converted</span> <span class="o">=</span> <span class="n">converted</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
                <span class="c1"># If converted == &quot;0&quot; will become &quot;&quot;</span>
                <span class="k">if</span> <span class="n">converted</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="n">converted</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
                <span class="n">converted</span> <span class="o">+=</span> <span class="n">letter</span>
                <span class="k">break</span>
        <span class="k">return</span> <span class="n">converted</span>

    <span class="k">def</span> <span class="nf">convert</span><span class="p">(</span><span class="nb">input</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert a quantity using metric prefixes.</span>

<span class="sd">        Args:</span>
<span class="sd">            input (int): The number to convert.</span>
<span class="sd">            decimal (int, optional): Number of decimals to use. Defaults to 2.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Converted number.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">si_prefixes</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="s2">&quot; p&quot;</span><span class="p">,</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="o">-</span><span class="mi">9</span><span class="p">,</span> <span class="o">-</span><span class="mi">12</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; n&quot;</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">9</span><span class="p">,</span> <span class="o">-</span><span class="mi">6</span><span class="p">,</span> <span class="o">-</span><span class="mi">9</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">6</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">6</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; m&quot;</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; k&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; M&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">6</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; G&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">9</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; T&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">12</span><span class="p">)),</span>
                <span class="p">(</span><span class="s2">&quot; P&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mi">15</span><span class="p">)),</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">input</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;0&quot;</span>

        <span class="n">input_log</span> <span class="o">=</span> <span class="n">log10</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
        <span class="c1"># i.e., -3.9 should become -4;</span>
        <span class="c1">#        3.9 should become 4</span>
        <span class="k">if</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">input_log</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">input_log</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">input_log</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ceil</span><span class="p">(</span><span class="n">input_log</span><span class="p">))</span>

        <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">letter</span><span class="p">,</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">si_prefixes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">min_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">max_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="n">max_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span>
            <span class="k">elif</span> <span class="n">min_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span> <span class="ow">and</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>

            <span class="k">if</span> <span class="n">in_interval</span><span class="p">:</span>
                <span class="n">converted</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">(</span><span class="nb">input</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">10</span><span class="o">**</span><span class="n">interval</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">:</span><span class="s2">.</span><span class="si">{</span><span class="n">decimal</span><span class="si">}</span><span class="s2">f</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="c1"># Not interested in any .0000 at the end</span>
                <span class="n">converted</span> <span class="o">=</span> <span class="n">converted</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
                <span class="c1"># If converted == &quot;0&quot; will become &quot;&quot;</span>
                <span class="k">if</span> <span class="n">converted</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="n">converted</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
                <span class="n">converted</span> <span class="o">+=</span> <span class="n">letter</span>
                <span class="k">break</span>
        <span class="k">return</span> <span class="n">converted</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="helix.utility.unit_prefix.UnitPrefix.convert" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">convert</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Convert a quantity using metric prefixes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>input</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number to convert.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>decimal</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of decimals to use. Defaults to 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Converted number.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>helix\utility\unit_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">convert</span><span class="p">(</span><span class="nb">input</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a quantity using metric prefixes.</span>

<span class="sd">    Args:</span>
<span class="sd">        input (int): The number to convert.</span>
<span class="sd">        decimal (int, optional): Number of decimals to use. Defaults to 2.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Converted number.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">si_prefixes</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot; p&quot;</span><span class="p">,</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="o">-</span><span class="mi">9</span><span class="p">,</span> <span class="o">-</span><span class="mi">12</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; n&quot;</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">9</span><span class="p">,</span> <span class="o">-</span><span class="mi">6</span><span class="p">,</span> <span class="o">-</span><span class="mi">9</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">6</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">6</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; m&quot;</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; k&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; M&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">6</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; G&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">9</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; T&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">12</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; P&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mi">15</span><span class="p">)),</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="nb">input</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;0&quot;</span>

    <span class="n">input_log</span> <span class="o">=</span> <span class="n">log10</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
    <span class="c1"># i.e., -3.9 should become -4;</span>
    <span class="c1">#        3.9 should become 4</span>
    <span class="k">if</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">input_log</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">input_log</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">input_log</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ceil</span><span class="p">(</span><span class="n">input_log</span><span class="p">))</span>

    <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">letter</span><span class="p">,</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">si_prefixes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">min_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">max_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">in_interval</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">max_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span>
        <span class="k">elif</span> <span class="n">min_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span> <span class="ow">and</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>

        <span class="k">if</span> <span class="n">in_interval</span><span class="p">:</span>
            <span class="n">converted</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">(</span><span class="nb">input</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">10</span><span class="o">**</span><span class="n">interval</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">:</span><span class="s2">.</span><span class="si">{</span><span class="n">decimal</span><span class="si">}</span><span class="s2">f</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="c1"># Not interested in any .0000 at the end</span>
            <span class="n">converted</span> <span class="o">=</span> <span class="n">converted</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
            <span class="c1"># If converted == &quot;0&quot; will become &quot;&quot;</span>
            <span class="k">if</span> <span class="n">converted</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">converted</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
            <span class="n">converted</span> <span class="o">+=</span> <span class="n">letter</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">converted</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="helix.utility.unit_prefix.UnitPrefix.convert_bytes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">convert_bytes</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

      <p>Convert bytes using binary prefixes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>input</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number to convert.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>decimal</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of decimals to use. Defaults to 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The input number is a float</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The input number is negative.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Converted number.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>helix\utility\unit_prefix.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">convert_bytes</span><span class="p">(</span><span class="nb">input</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert bytes using binary prefixes.</span>

<span class="sd">    Args:</span>
<span class="sd">        input (int): The number to convert.</span>
<span class="sd">        decimal (int, optional): Number of decimals to use. Defaults to 2.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: The input number is a float</span>
<span class="sd">        ValueError: The input number is negative.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Converted number.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">input</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Can&#39;t convert a fractional number of bytes&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">input</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Can&#39;t convert a negative number of bytes&quot;</span><span class="p">)</span>
    <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">si_prefixes</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="s2">&quot; B&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">0</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; kB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; MB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">20</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; GB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">30</span><span class="p">)),</span>
            <span class="p">(</span><span class="s2">&quot; TB&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mi">40</span><span class="p">)),</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="nb">input</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;0 B&quot;</span>

    <span class="n">input_log</span> <span class="o">=</span> <span class="n">log</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">input_log</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">floor</span><span class="p">(</span><span class="n">input_log</span><span class="p">))</span>

    <span class="n">converted</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">letter</span><span class="p">,</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">si_prefixes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">min_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">max_interval</span> <span class="o">=</span> <span class="n">interval</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">in_interval</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">max_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span>
        <span class="k">elif</span> <span class="n">min_interval</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">in_interval</span> <span class="o">=</span> <span class="n">input_log</span> <span class="o">&gt;=</span> <span class="n">min_interval</span> <span class="ow">and</span> <span class="n">input_log</span> <span class="o">&lt;</span> <span class="n">max_interval</span>

        <span class="k">if</span> <span class="n">in_interval</span><span class="p">:</span>
            <span class="n">converted</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">(</span><span class="nb">input</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">2</span><span class="o">**</span><span class="n">interval</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">:</span><span class="s2">.</span><span class="si">{</span><span class="n">decimal</span><span class="si">}</span><span class="s2">f</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="c1"># Not interested in any .0000 at the end</span>
            <span class="n">converted</span> <span class="o">=</span> <span class="n">converted</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
            <span class="c1"># If converted == &quot;0&quot; will become &quot;&quot;</span>
            <span class="k">if</span> <span class="n">converted</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">converted</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
            <span class="n">converted</span> <span class="o">+=</span> <span class="n">letter</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">converted</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
    
  </body>
</html>